### PR TITLE
scheduler query component load tracking

### PR DESCRIPTION
**Pull request by francoposa**
#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/scheduler/queue/query_component_load.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	if len(req.AdditionalQueueDimensions) > 0 {
+		expectedQueryComponent = req.AdditionalQueueDimensions[0]
+	}
+
+	return expectedQueryComponent
+}
+
+// queryComponentFlags interprets annotations by the frontend for the expected query component,
+// and flags whether a query request is expected to be served by the ingesters, store-gateways, or both.
+func queryComponentFlags(queryComponentName string) (isIngester, isStoreGateway bool) {
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_load.go:44]: If I understand this correctly, this takes in a queue dimension (`ingester`, `store-gateway` or `ingester-and-store-gateway`), not a single component name (`ingester` or `store-gateway`). Perhaps `queueDimension` would be a clearer name rather than `queryComponentName`?

File: pkg/scheduler/queue/query_component_load.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+var QueryComponents = []QueryComponent{
+	StoreGateway,
+	Ingester,
+}
+
+// cannot import constants from frontend/v2 due to import cycle
+// these are attached to the request's AdditionalQueueDimensions by the frontend.
+const ingesterQueueDimension = "ingester"
+const storeGatewayQueueDimension = "store-gateway"
+const ingesterAndStoreGatewayQueueDimension = "ingester-and-store-gateway"
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_load.go:30]: Perhaps we should create a `QueueDimension` type like you have for `QueryComponent` to make it clearer when we expect one or the other?

File: pkg/scheduler/queue/query_component_load.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+// overloadThreshold calculates the minimum number of inflight requests to qualify a query component as overloaded.
+// This is determined by (inflight requests for second-most-loaded component) * (overload factor).
+// Intended to only be called within a read lock on inflightRequestsMu.
+func (qcl *QueryComponentLoad) overloadThreshold() int {
+	// overload factor is expected to be > 1; any misconfiguration of this should be a no-op
+	if qcl.overloadFactor <= 1 {
+		return 0
+	}
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_load.go:134]: Is this necessary given the validation of `overloadFactor` in `NewQueryComponentLoad`?

File: pkg/scheduler/queue/query_component_load_test.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+	for i := 0; i < requestCount; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			<-start
+			mockForwardRequestToQuerier(t, queryComponentLoad)
+		}()
+	}
+	close(start)
+	wg.Wait()
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_load_test.go:51]: Might be worth adding an assertion here to confirm that the inflight request count is 0 after all the incrementing and decrementing - this would give us extra confidence that there are no concurrency issues.

File: pkg/scheduler/queue/query_component_load_test.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+		t.Run(testCase.name, func(t *testing.T) {
+			queryComponentLoad, err := NewQueryComponentLoad(testOverloadFactor, prometheus.NewPedanticRegistry())
+			require.NoError(t, err)
+
+			alwaysNotOverloaded := &QueryComponentLoad{
+				querierInflightRequestsByComponent: make(map[QueryComponent]int),
+				querierInflightRequestsTotal:       0,
+				// constructor disallows anything <=1; set it manually here
+				overloadFactor: rand.Float64(),
+			}
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_load_test.go:199]: It's not clear to me what this is exercising - a comment here might be handy.

File: pkg/scheduler/queue/query_component_load.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+// cannot import constants from frontend/v2 due to import cycle
+// these are attached to the request's AdditionalQueueDimensions by the frontend.
+const ingesterQueueDimension = "ingester"
+const storeGatewayQueueDimension = "store-gateway"
+const ingesterAndStoreGatewayQueueDimension = "ingester-and-store-gateway"
+
+// QueryComponentNameForRequest parses the expected query component from annotations by the frontend.
+func QueryComponentNameForRequest(req *SchedulerRequest) string {
```
[the remark this answers is not in the document]
Comment (francoposa) [reply pkg/scheduler/queue/query_component_load.go:33]: I can add a comment for now - ideally now that we moved away from this idea of the queue structure as hyper flexible / extensible structure, I would similarly like to move away from "Additional Queue Dimensions" as it doesn't make sense any more.

This concept of "Expected Query Component" can get a proper name and be put into the `Hints` on the RPC message, rather than continuing this dance of "the expected query component is the first item of the additional queue dimensions array, also there's no other use for this array".

File: pkg/scheduler/queue/query_component_load.go (hunk 1 above)
Comment (francoposa) [reply pkg/scheduler/queue/query_component_load.go:44]: same idea as above - I want this `queueDimension` concept to go away because it's not accurate or really useful.

File: pkg/scheduler/queue/query_component_load.go (hunk 2 above)
Comment (francoposa) [reply pkg/scheduler/queue/query_component_load.go:30]: see above

File: pkg/scheduler/queue/query_component_capacity.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+		storeGatewayInflightRequests: atomic.NewInt64(0),
+		querierInflightRequestsTotal: atomic.NewInt64(0),
+
+		querierInflightRequestsGauge: querierInflightRequests,
+	}, nil
+}
+
+// ExceedsCapacityForComponentName checks whether a query component has exceeded the capacity utilization threshold.
+// This enables the dequeuing algorithm to skip requests for a query component experiencing heavy load,
+// reserving querier-worker connection capacity to continue servicing requests for the other component.
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_capacity.go:102]: I know you haven't implemented it yet, but this isn't quite how I imagined the dequeuing working.

tl;dr: querier workers should never be idle if there are requests in the queue, but this comment makes it sound like we would keep some workers idle

Rather than skipping requests for a query component experiencing heavy load and reserving capacity for requests for the other component, I think we should instead just prioritise requests for the component under low load. This way, if (say) store-gateways are under high load, but there are no waiting requests for ingesters, we'll still make use of the available querier capacity to work on store-gateway requests. But, if ingester requests later arrive, we'd then pick those to work on next over other older requests for store-gateways.

Comment (charleskorn) [inline pkg/scheduler/queue/query_component_capacity.go:102]: I also think we need to track three kinds of queries, rather than whether they touch ingesters or store-gateways:
* ingester-only
* store-gateway-only
* ingester and store-gateway

The reason for this is that otherwise, there's a risk that requests that touch both components would starve requests that touch only one. (For example, imagine a scenario where store-gateways are running slowly - all of the capacity we'd like to allocate to ingester-only queries could be consumed by queries that touch both ingesters and store-gateways.)

Comment (francoposa) [reply pkg/scheduler/queue/query_component_capacity.go:102]: this is will be handled by the queue structure that Casie is working on, using this method. this method gives that exact information - if something touches both and either one has exceeded the threshold it will let you know

Comment (charleskorn) [reply pkg/scheduler/queue/query_component_capacity.go:102]: I think I'm missing something - how does this method track "ingester+store-gateway" requests separately from "ingester-only" requests?

Comment (francoposa) [reply pkg/scheduler/queue/query_component_capacity.go:102]: It's not this components job to "track" anything about the requests or queue at all.
It is simply going to tell the other component (being worked on by Casie) that navigates through the queue structure whether any of the query components (ingester or store gateway) exceeds the reserved capacity _for that expected component name_, where the expected component name is expected to be `ingester`, `store-gateway`, `ingester-and-store-gateway`, or other (which is treated as ingester and store gateway).

`isIngester, isStoreGateway := queryComponentFlags(name)`
both of those flags will be true if the expected component name is anything other than just "ingester" or just "store-gateway".

When both flags are true, the capacity utliization is checked for both.
that's the `if isIngester { // return if load exceeded }` and `if isStoreGateway { // return if load exceeded }` parts of this method.

File: pkg/scheduler/queue/query_component_capacity.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+)
+
+// cannot import constants from frontend/v2 due to import cycle
+// these are attached to the request's AdditionalQueueDimensions by the frontend.
+const ingesterQueueDimension = "ingester"
+const storeGatewayQueueDimension = "store-gateway"
+const ingesterAndStoreGatewayQueueDimension = "ingester-and-store-gateway"
+
+// QueryComponentNameForRequest parses the expected query component from annotations by the frontend.
+func QueryComponentNameForRequest(req *SchedulerRequest) string {
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_capacity.go:27]: I find the name "component name" very confusing given it could contain multiple components. Perhaps "affected components" would be clearer? 

(same feedback applies for `ExceedsCapacityForComponentName` below - I think something like `ExceedsCapacityForAffectedComponents` might be clearer)

File: pkg/scheduler/queue/query_component_capacity.go [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+// A QueryComponent's utilization is allowed to exceed the reserved capacity when the queue is not backlogged.
+// If an influx of queries then creates a backlog, this method will indicate to skip queries for the component.
+// As the inflight queries complete or fail, the component's utilization will naturally decrease.
+// This method will continue to indicate to skip queries for the component until it is back under the threshold.
+func (qcl *QueryComponentCapacity) ExceedsCapacityForComponentName(
+	name string, connectedWorkers *atomic.Int64, queueLen, waitingWorkers int,
+) (bool, QueryComponent) {
+	if waitingWorkers > queueLen {
+		// excess querier-worker capacity; no need to reserve any for now
+		return false, ""
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_capacity.go:129]: Do we need both the boolean and string return value here? Seems like we just need to know whether we should or shouldn't avoid this query request.

File: pkg/scheduler/queue/query_component_capacity_test.go [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+			queryComponentName:           ingesterQueueDimension,
+			thresholdExceededComponent:   "",
+		},
+		{
+			name:                         "ingester and store-gateway, no backlog, store gateway utilization above threshold",
+			queueLen:                     2,
+			waitingWorkers:               3,
+			ingesterInflightRequests:     0,
+			storegatewayInflightRequests: 7,
+			queryComponentName:           storeGatewayQueueDimension,
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_capacity_test.go:112]: Should these both use `ingesterAndStoreGatewayQueueDimension`?

File: pkg/scheduler/queue/query_component_capacity_test.go [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+	// no component can have more than 6 inflight requests.
+	testReservedCapacity := 0.4
+
+	testCases := []struct {
+		name string
+
+		queueLen                     int
+		waitingWorkers               int
+		ingesterInflightRequests     int
+		storegatewayInflightRequests int
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_capacity_test.go:71]: [nit]

```suggestion
		storeGatewayInflightRequests int
```

File: pkg/scheduler/queue/queue.go [hunk 12]
```diff
@@ -291,6 +302,29 @@ func (q *RequestQueue) trySendNextRequestForQuerier(call *nextRequestForQuerierC
...
+			exceedsCapacity, queryComponent := q.QueryComponentCapacity.ExceedsCapacityForComponentName(
+				queryComponentName,
+				q.connectedQuerierWorkers,
+				q.queueBroker.tenantQueuesTree.ItemCount(),
+				q.waitingNextRequestForQuerierCalls.Len(),
+			)
+
+			if exceedsCapacity {
+				level.Warn(q.log).Log(
+					"msg", "querier worker connections in use by query component exceed reserve capacity",
```
Comment (charleskorn) [inline pkg/scheduler/queue/queue.go:319]: Should we make this clearer that this is just for testing and the query-scheduler didn't do anything different as a result of this situation?

(I'm imagining someone being in the heat of an incident, seeing this message and going off on a wild goose chase, not realising this is just for evaluation purposes.)

Given this, we should probably also drop the severity down to debug as well.

File: pkg/scheduler/scheduler.go [hunk 13]
```diff
@@ -148,7 +150,26 @@ func NewScheduler(cfg Config, limits Limits, log log.Logger, registerer promethe
 		Name: "cortex_query_scheduler_enqueue_duration_seconds",
 		Help: "Time spent by requests waiting to join the queue or be rejected.",
 	})
-	s.requestQueue = queue.NewRequestQueue(s.log, cfg.MaxOutstandingPerTenant, cfg.AdditionalQueryQueueDimensionsEnabled, cfg.QuerierForgetDelay, s.queueLength, s.discardedRequests, enqueueDuration)
+	querierInflightRequestsGauge := promauto.With(registerer).NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "cortex_query_scheduler_querier_inflight_requests",
+			Help: "Number of inflight requests being processed on a querier-scheduler connection.",
```
Comment (charleskorn) [inline pkg/scheduler/scheduler.go:156]: [nit] This is clearer to me:

```suggestion
			Help: "Number of inflight requests being processed on all querier-scheduler connections.",
```

File: pkg/scheduler/queue/query_component_capacity.go [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+	querierInflightRequestsTotal *atomic.Int64
+
+	querierInflightRequestsGauge *prometheus.GaugeVec
+}
+
+// DefaultReservedQueryComponentCapacity reserves 1 / 3 of querier-worker connections
+// for the query component utilizing fewer of the available connections.
+const DefaultReservedQueryComponentCapacity = 0.33
+
+const MinReservedQueryComponentCapacity = 0.1
```
[the remark this answers is not in the document]
Comment (francoposa) [reply pkg/scheduler/queue/query_component_capacity.go:77]: Honestly I did not have a good reason for 10% other than it didn't seem all that useful to reserve much less than this
If the point is to keep queries for one component moving when the other component is struggling, it seems like a waste to allow the struggling component to dominate the capacity. This would cap it at 90% at least.

We could allow it to be down to zero, which would be like disabling the feature. That could make sense if we plan to make this a configurable feature in the future

[the remark this answers is not in the document]
Comment (charleskorn) [reply pkg/scheduler/queue/query_component_capacity.go:77]: I think having a way to disable this would be valuable, especially while we're still experimenting with it.

File: pkg/scheduler/queue/query_component_capacity.go (hunk 8 above)
Comment (francoposa) [reply pkg/scheduler/queue/query_component_capacity.go:27]: The fundamental problem is that we have a single string that represents both being passed from the front end, and we have to call that something.

We also need a name for the separate components themselves, as well as these flags that translate from the single-string component name to when a given component used, like `isStoreGateway` and `isIngester`.

I don't use "Components" because that seems to indicate multiple of our type, `QueryComponent`.

I have tried to disambiguate a bit here by using:
* single string: `ExpectedQueryComponentName` / `ComponentName`
* flags: `Flags`
* individual components: the `QueryComponent` type

I think ultimately the solutino here is the solution for everything else, which is fix it in the frontend.

We can abandon this:
`const ingesterAndStoreGatewayQueueDimension = "ingester-and-store-gateway"`

for just something like

```golang
type ExpectedQueryComponents struct {
    ingester bool
    storeGateway bool
}
```
then this whole concept of "something that is a single value but indicates multiple things" is gone

File: pkg/scheduler/queue/query_component_capacity.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (francoposa) [reply pkg/scheduler/queue/query_component_capacity.go:77]: Set to not have a minumum

File: pkg/scheduler/queue/query_component_capacity.go [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+		ingesterInflightRequests:     atomic.NewInt64(0),
+		storeGatewayInflightRequests: atomic.NewInt64(0),
+		querierInflightRequestsTotal: atomic.NewInt64(0),
+
+		querierInflightRequestsGauge: querierInflightRequests,
+	}, nil
+}
+
+// ExceedsCapacityForComponentName checks whether a query component has exceeded the capacity utilization threshold.
```
[the remark this answers is not in the document]
Comment (francoposa) [reply pkg/scheduler/queue/query_component_capacity.go:100]: Capacity utilization threshold is correct. 
Capacity utilization is what it is using, as defined in the docstring. Capacity utilization threshold is the max it's allowed to use

Capacity reservation is what it's _not_ allowed to use.

File: pkg/scheduler/queue/query_component_capacity.go [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+//
+// Capacity reservation only occurs when the queue backlogged, where backlogged is defined as
+// (length of the query queue) >= (number of querier-worker connections waiting for a query).
+//
+// A QueryComponent's utilization is allowed to exceed the reserved capacity when the queue is not backlogged.
+// If an influx of queries then creates a backlog, this method will indicate to skip queries for the component.
+// As the inflight queries complete or fail, the component's utilization will naturally decrease.
+// This method will continue to indicate to skip queries for the component until it is back under the threshold.
+func (qcl *QueryComponentCapacity) ExceedsCapacityForComponentName(
+	name string, connectedWorkers *atomic.Int64, queueLen, waitingWorkers int,
```
[the remark this answers is not in the document]
Comment (francoposa) [reply pkg/scheduler/queue/query_component_capacity.go:125]: :+1: done

File: pkg/scheduler/queue/query_component_capacity.go (hunk 9 above)
Comment (francoposa) [reply pkg/scheduler/queue/query_component_capacity.go:129]: it's useful for tests and could be included in the logging if we wanted.

File: pkg/scheduler/queue/query_component_capacity.go [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+func (qcl *QueryComponentCapacity) IncrementForComponentName(expectedQueryComponent string) {
+	qcl.updateForComponentName(expectedQueryComponent, 1)
+}
+
+// DecrementForComponentName is called when a querier completes or fails a request
+func (qcl *QueryComponentCapacity) DecrementForComponentName(expectedQueryComponent string) {
+	qcl.updateForComponentName(expectedQueryComponent, -1)
+}
+
+func (qcl *QueryComponentCapacity) updateForComponentName(expectedQueryComponent string, increment int64) {
```
[the remark this answers is not in the document]
Comment (francoposa) [reply pkg/scheduler/queue/query_component_capacity.go:167]: I try to use `expectedQueryComponent` everywhere because it does not tell the querier which query component to call for, it just tries to predict it.

It should be very rare, but the threshold for query-ingesters-within or query-store-after could be crossed after this gets annotated in the front end and before they query is actually executed, making it inaccurate.

File: pkg/scheduler/queue/queue.go (hunk 12 above)
Comment (francoposa) [reply pkg/scheduler/queue/queue.go]: updated the message and changed to Info level. We don't want debug level because then we have to turn debug logging on to observe it. Info works because it's not warn or error, which would be filtered for when investigating an incident.

File: pkg/scheduler/queue/query_component_capacity.go (hunk 8 above)
Comment (charleskorn) [reply pkg/scheduler/queue/query_component_capacity.go]: >I think ultimately the solution here is the solution for everything else, which is fix it in the frontend.

That sounds good to me, but I'm happy for that to come in a separate PR.

File: pkg/scheduler/queue/query_component_capacity.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (charleskorn) [reply pkg/scheduler/queue/query_component_capacity.go:162]: Right, makes sense, thanks for clarifying.

File: pkg/scheduler/queue/query_component_capacity_test.go [hunk 18]
```diff
@@ -241,13 +241,13 @@ func TestExceedsReservedCapacityForQueryComponents(t *testing.T) {
...
 
-			for i := 0; i < testCase.storegatewayInflightRequests; i++ {
+			for i := 0; i < testCase.storeGatewayInflightRequests; i++ {
 				queryComponentLoad.IncrementForComponentName(storeGatewayQueueDimension)
 			}
 
 			exceedsCapacity, queryComponent := queryComponentLoad.ExceedsCapacityForComponentName(
 				testCase.queryComponentName,
-				connectedWorkers,
+				connectedWorkers.Load(),
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_capacity_test.go:250]: Can't `connectedWorkers` be an ordinary (non-atomic) integer now?

File: pkg/scheduler/scheduler.go [hunk 19]
```diff
@@ -450,6 +471,10 @@ func (s *Scheduler) forwardRequestToQuerier(querier schedulerpb.SchedulerForQuer
 	// Make sure to cancel request at the end to clean up resources.
 	defer s.cancelRequestAndRemoveFromPending(req.FrontendAddress, req.QueryID, "request complete")
 
+	queryComponentName := req.ExpectedQueryComponentName()
+	s.requestQueue.QueryComponentCapacity.IncrementForComponentName(queryComponentName)
```
Comment (charleskorn) [inline pkg/scheduler/scheduler.go:475]: Calling `IncrementForComponentName` here introduces a race: we'll make a decision based on the current utilisation in `RequestQueue.trySendNextRequestForQuerier`, but only later update the utilisation here. In the meantime, we may have made more decisions for other requests without accounting for this query.

I think this has to be done synchronously as part of `RequestQueue`'s dispatcher loop.

File: pkg/scheduler/scheduler.go [hunk 20]
```diff
@@ -450,6 +471,10 @@ func (s *Scheduler) forwardRequestToQuerier(querier schedulerpb.SchedulerForQuer
...
+	defer s.requestQueue.QueryComponentCapacity.DecrementForComponentName(queryComponentName)
```
Comment (charleskorn) [inline pkg/scheduler/scheduler.go:476]: For symmetry with the above change (moving the call to `IncrementForComponentName` into the dispatcher loop), what do you think about changing this to a call that is handled by `RequestQueue`'s dispatcher loop as well? 

This would then also mean `QueryComponentCapacity` would no longer have to deal with concurrency at all, as every interaction with it will come from the dispatcher loop.

File: pkg/scheduler/scheduler.go (hunk 19 above)
Comment (charleskorn) [inline pkg/scheduler/scheduler.go:475]: [nit] `QueryComponentCapacity.IncrementForComponentName` makes it sound like we're incrementing the capacity, but we're incrementing the inflight requests or utilisation. What do you think about renaming either `QueryComponentCapacity` or `IncrementForComponentName` to make this clearer? (eg. `IncrementInflightRequestsForComponentName`)

File: pkg/scheduler/queue/query_component_capacity.go [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+		return false, ""
+	}
+	if connectedWorkers <= 1 {
+		// corner case; cannot reserve capacity with only one worker available
+		return false, ""
+	}
+
+	// allow the functionality to be turned off via setting targetReservedCapacity to 0
+	minReservedConnections := int64(0)
+	if qcl.targetReservedCapacity > 0 {
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_capacity.go:129]: It'd be good to add a test for this case.

File: pkg/scheduler/queue/query_component_capacity.go [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+func (qcl *QueryComponentCapacity) ExceedsCapacityForComponentName(
+	name string, connectedWorkers int64, queueLen, waitingWorkers int,
+) (bool, QueryComponent) {
+	if waitingWorkers > queueLen {
+		// excess querier-worker capacity; no need to reserve any for now
+		return false, ""
+	}
+	if connectedWorkers <= 1 {
+		// corner case; cannot reserve capacity with only one worker available
+		return false, ""
```
Comment (charleskorn) [inline pkg/scheduler/queue/query_component_capacity.go:124]: It'd be good to add a test for this case as well.

File: pkg/scheduler/scheduler.go (hunk 19 above)
Comment (francoposa) [reply pkg/scheduler/scheduler.go:475]: renamed QueryComponentCapacity to `QueryComponentUtilization`, that way the utilization is what we are incrementing and decrementing.

Already had a bunch the "utilization" language in dosctrings and comments, I think it's fairly clear.

File: pkg/scheduler/scheduler.go (hunk 20 above)
Comment (francoposa) [reply pkg/scheduler/scheduler.go:476]: marking down from slack discussion:

the decrementing has to happen in the scheduler because that's the only component that knows when the request finishes.

if we want to both in the request queue we’d have to introduce another channel or similar communication layer to tell the request queue it completed so the queue can call decrement, but that would be out of scope for this PR.

File: pkg/scheduler/scheduler.go (hunk 19 above)
Comment (francoposa) [reply pkg/scheduler/scheduler.go:475]: from slack discussion: will add comment for now, as the RequestQueue is not ready to utilize this

Comment (francoposa) [reply pkg/scheduler/scheduler.go:475]: moved

**Review by charleskorn — APPROVED** [review]
Changes look good to me, I'm happy for this to be merged once these two comments (adding further test cases) are addressed:

- https://github.com/grafana/mimir/pull/8132#discussion_r1610917680
- https://github.com/grafana/mimir/pull/8132#discussion_r1610917801

---

Summarising the changes we've discussed for a future PR:

- [Moving responsibility for calling `DecrementForComponentName` to `RequestQueue` and removing use of atomics in `QueryComponentUtilization`](https://github.com/grafana/mimir/pull/8132#discussion_r1610916255)
- [Passing the expected affected query components as another type embedded in the request hints, rather than as a string](https://github.com/grafana/mimir/pull/8132#discussion_r1610544613)
